### PR TITLE
change default directories

### PR DESCRIPTION
**Pull request by borg323**
Does the lc0 part of #976.

### CODE REVIEW

**Comment by borg323** [comment]
What this does:
1. For network file search with `<autodiscover>`, the following directories are searched in order:
- `.` and `networks`
- `lc0` and `lc0/networks` under `$XDG_DATA_HOME` (or `~/.share/local` if not set)
- `lc0` and `lc0/networks` under every path in `$XDG_DATA_DIRS` (or `/usr/local/share/:/usr/share/` if not set)
2. For the config file by default the following directories are searched in order for `lc0.config`:
- `.`
- `lc0` under `$XDG_CONFIG_HOME` (or `~/.config` if not set)
- `lc0` under every path in `$XDG_CONFIG_DIRS` (or ` /etc/xdg` if not set)
Note that currently the help text shows `lc0.config` but `<default>` is used internally. This is not ideal, maybe we should change the help text to `<search for lc0.config>`.
3. Training game generation is done in `lc0/data-<random>` under `$XDG_CACHE_HOME` (or `~/.cache` if not set).

4. The `leelaz_opencl_tuning` file is searched and saved in `lc0` under `$XDG_CACHE_HOME` (or `~/.cache` if not set), with an already existing backend option to override.

**Comment by mooskagh** [comment]
That will probably mean that older/newer client won't be compatible with newer/older lc0?
Because lc0 writes files not where the client expects them.
Probably not a problem, but would be good to know whether that's the case.

**Comment by borg323** [comment]
>  Because lc0 writes files not where the client expects them.

The client gets the path by parsing lc0 output, this works fine (tested). The opposite may not be true, with new client running old lc0 from the path, but this is a new use case, so requiring new lc0 seems reasonable.

File: src/neural/opencl/network_opencl.cc [hunk 1]
```diff
@@ -245,8 +246,11 @@ class OpenCLNetwork : public Network {
     params_.tune_only = options.GetOrDefault<bool>("tune_only", false);
     params_.tune_exhaustive =
         options.GetOrDefault<bool>("tune_exhaustive", false);
-    params_.tuner_file =
-        options.GetOrDefault<std::string>("tuner_file", "leelaz_opencl_tuning");
+    if (options.IsDefault<std::string>("tuner_file")) {
+      params_.tuner_file = GetUserCacheDirectory() + "lc0/leelaz_opencl_tuning";
```
Comment (Tilps) [inline src/neural/opencl/network_opencl.cc:250]: This means multiple installs of lc0 get the same tuner file unless you use backend-opts to override?
I wonder if that will ever cause us pain...

Comment (Tilps) [reply src/neural/opencl/network_opencl.cc:250]: hmm only on linux it seems - windows this is empty string, so maybe I am fine with that :P

File: src/neural/writer.cc [hunk 2]
```diff
@@ -36,9 +36,20 @@
...
+}
+
+}  // namespace
+
 TrainingDataWriter::TrainingDataWriter(int game_id) {
   static std::string directory =
-      GetUserCacheDirectory() + "lc0/data-" + Random::Get().GetString(12);
+      GetLc0CacheDirectory() + "data-" + Random::Get().GetString(12);
   // It's fine if it already exists.
   CreateDirectory(directory.c_str());
```
Comment (Tilps) [inline src/neural/writer.cc:54]: This needs to be recursive create.

File: src/neural/opencl/network_opencl.cc [hunk 3]
```diff
@@ -247,7 +247,11 @@ class OpenCLNetwork : public Network {
     params_.tune_exhaustive =
         options.GetOrDefault<bool>("tune_exhaustive", false);
     if (options.IsDefault<std::string>("tuner_file")) {
-      params_.tuner_file = GetUserCacheDirectory() + "lc0/leelaz_opencl_tuning";
+      std::string user_cache_path = GetUserCacheDirectory();
+      if (!user_cache_path.empty()) {
+        user_cache_path += "lc0/";
+      }
+      params_.tuner_file = user_cache_path + "leelaz_opencl_tuning";
```
Comment (Tilps) [inline src/neural/opencl/network_opencl.cc:254]: If user_cache_path is not empty - it needs a CreateDirectory before this.

**Pull request merged**